packages/transformers/js/core/src/side_effects.rs

```diff
@@ -0,0 +1,714 @@
+use std::collections::HashSet;
+#[cfg(test)]
+use std::path::PathBuf;
+use swc_core::common::comments::SingleThreadedComments;
+#[cfg(test)]
+use swc_core::common::sync::Lrc;
+#[cfg(test)]
+use swc_core::common::{FileName, SourceMap};
+use swc_core::ecma::ast::*;
+use swc_core::ecma::atoms::{js_word, JsWord};
+#[cfg(test)]
+use swc_core::ecma::parser::lexer::Lexer;
+#[cfg(test)]
+use swc_core::ecma::parser::{EsConfig, Parser, StringInput, Syntax};
+use swc_core::ecma::visit::{Visit, VisitWith};
+
+use crate::hoist::ImportedSymbol;
+use crate::id;
+use crate::utils::{match_member_expr, match_str};
+
+pub struct SideEffects<'a> {
+  pub has_side_effects: bool,
+  decls: HashSet<Id>,
+  imports: HashSet<JsWord>,
+  comments: &'a SingleThreadedComments,
+}
+
+impl<'a> SideEffects<'a> {
+  pub fn new(
+    decls: HashSet<Id>,
+    imported_symbols: &Vec<ImportedSymbol>,
+    comments: &'a SingleThreadedComments,
+  ) -> Self {
+    let imports = HashSet::from_iter(
+      imported_symbols
+        .into_iter()
+        .map(|symbol| symbol.local.clone()),
+    );
+
+    SideEffects {
+      has_side_effects: false,
+      decls,
+      imports,
+      comments,
+    }
+  }
+
+  // TODO use utils::match_require instead
+  fn match_require(&self, node: &Expr) -> Option<JsWord> {
+    match node {
+      Expr::Call(call) => match &call.callee {
+        Callee::Expr(expr) => match &**expr {
+          Expr::Ident(ident) => {
+            if ident.sym == js_word!("require")
+              && !self.decls.contains(&(ident.sym.clone(), ident.span.ctxt))
+            {
+              if let Some(arg) = call.args.get(0) {
+                return match_str(&arg.expr).map(|(name, _)| name);
+              }
+            }
+
+            None
+          }
+          Expr::Member(member) => {
+            if match_member_expr(member, vec!["module", "require"], &self.decls) {
+              if let Some(arg) = call.args.get(0) {
+                return match_str(&arg.expr).map(|(name, _)| name);
+              }
+            }
+
+            None
+          }
+          _ => None,
+        },
+        _ => None,
+      },
+      _ => None,
+    }
+  }
+
+  fn is_safe_variable_assignment(&self, node: &PatOrExpr) -> bool {
+    if let Some(ident) = node.as_ident() {
+      return self.decls.contains(&id!(ident)) && !self.imports.contains(&ident.sym);
+    }
+
+    if let Some(expr) = node.as_expr() {
+      match expr {
+        Expr::Member(member_expr) => {
+          if match_member_expr(&member_expr, vec!["module", "exports"], &self.decls)
```

Comment:
Is this basically saying that top level assignments to `module.<exports|hot|require>` are safe?

.. and the below is that only access of previously declared variables or imports are safe?

Reply:
> Is this basically saying that top level assignments to module.<exports|hot|require> are safe?
Yes

> and the below is that only access of previously declared variables or imports are safe?
Not access but assignment. Also I only allow assignment of locals declared in the file. I don't allow assignment to imports as that would mean the current asset is mutating an import in the top level scope which is a side effect.